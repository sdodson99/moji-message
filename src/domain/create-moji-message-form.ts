export type CreateMojiMessageForm = {
  message?: string;
  messageEmoji?: string;
  backgroundEmoji?: string;
};
