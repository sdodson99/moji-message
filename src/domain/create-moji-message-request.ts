export type CreateMojiMessageRequest = {
  message: string;
  messageEmoji: string;
  backgroundEmoji: string;
};
