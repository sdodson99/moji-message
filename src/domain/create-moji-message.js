import { CharacterMatrixMap } from './character-matrix-map';

const CHARACTER_GAP = '\n\n\n';
const EMPTY_CELL_GAP = '     ';

/**
 * Convert a message to a Moji Message.
 * @param {string} message The message to convert.
 * @param {string} emoji The emoji to build the Moji Message from.
 * @returns The converted Moji Message.
 */
export function createMojiMessage(message, emoji) {
  const messageCharacters = [...message];

  const mojiMessageCharacters = messageCharacters.map((c) => createMojiCharacter(c, emoji));

  const mojiMessage = mojiMessageCharacters.join(CHARACTER_GAP);

  return mojiMessage;
}

/**
 * Convert a message character to a Moji Message character
 * @param {string} character A single character to convert.
 * @param {string} emoji The emoji to build the Moji Message character from.
 * @returns The converted Moji Message character.
 */
function createMojiCharacter(character, emoji) {
  const upperCaseCharacter = character.toUpperCase();

  const characterMatrix = CharacterMatrixMap[upperCaseCharacter];

  if (!characterMatrix) {
    return '';
  }

  const moijCharacterRows = characterMatrix.map((row) => createMojiCharacterRow(row, emoji));

  return moijCharacterRows.join('\n');
}

/**
 * Convert a character matrix row to the appropiate emoji representation.
 * @param {number[]} characterMatrixRow The row of character matrix cells.
 * @param {string} emoji The emoji to convert the character matrix to.
 * @returns The converted character matrix row value.
 */
function createMojiCharacterRow(characterMatrixRow, emoji) {
  const mojiCharacterCells = characterMatrixRow.map((cell) => createMojiCharacterCell(cell, emoji));

  return mojiCharacterCells.join('');
}

/**
 * Convert a character matrix cell to the appropiate emoji representation.
 * @param {number} characterMatrixCell The binary value in the character matrix cell.
 * @param {string} emoji The emoji to convert the character matrix to.
 * @returns The converted character matrix cell value.
 */
function createMojiCharacterCell(characterMatrixCell, emoji) {
  if (!characterMatrixCell) {
    return EMPTY_CELL_GAP;
  }

  return emoji;
}
