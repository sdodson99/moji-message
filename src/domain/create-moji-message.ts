import { CharacterMatrixMap } from './character-matrix-map';

export function createMojiMessage(
  message: string,
  messageEmoji: string,
  backgroundEmoji: string
) {
  const messageCharacters = [...message];

  const mojiMessageCharacters = messageCharacters.map((c) =>
    createMojiCharacter(c, messageEmoji, backgroundEmoji)
  );

  const mojiMessage = mojiMessageCharacters.join(
    createCharacterGap(backgroundEmoji)
  );

  return mojiMessage;
}

function createMojiCharacter(
  character: string,
  messageEmoji: string,
  backgroundEmoji: string
) {
  const upperCaseCharacter = character.toUpperCase();

  const characterMatrix = CharacterMatrixMap[upperCaseCharacter];

  if (!characterMatrix) {
    return '';
  }

  const moijCharacterRows = characterMatrix.map((row) =>
    createMojiCharacterRow(row, messageEmoji, backgroundEmoji)
  );

  return moijCharacterRows.join('\n');
}

function createMojiCharacterRow(
  characterMatrixRow: boolean[],
  messageEmoji: string,
  backgroundEmoji: string
) {
  const mojiCharacterCells = characterMatrixRow.map((cell) =>
    createMojiCharacterCell(cell, messageEmoji, backgroundEmoji)
  );

  return mojiCharacterCells.join('');
}

function createMojiCharacterCell(
  characterMatrixCell: boolean,
  messageEmoji: string,
  backgroundEmoji: string
) {
  if (!characterMatrixCell) {
    return backgroundEmoji;
  }

  return messageEmoji;
}

function createCharacterGap(emoji: string) {
  return `\n${emoji.repeat(5)}\n`;
}
