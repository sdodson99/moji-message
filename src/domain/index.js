export * from './create-moji-message';
