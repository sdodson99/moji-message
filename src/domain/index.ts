export * from './create-moji-message';
export * from './create-moji-message-request';
export * from './create-moji-message-store';
