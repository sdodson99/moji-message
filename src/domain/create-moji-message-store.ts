import { CreateMojiMessageRequest } from './create-moji-message-request';

export class CreateMojiMessageStore {
  private _currentMojiMessageRequest?: CreateMojiMessageRequest;

  get currentMojiMessageRequest() {
    return this._currentMojiMessageRequest;
  }

  set currentMojiMessageRequest(value) {
    this._currentMojiMessageRequest = value;
  }
}
