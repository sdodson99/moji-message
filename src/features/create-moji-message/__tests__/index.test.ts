import type { PopupPickerController } from '@picmo/popup-picker';
import {
  CreateMojiMessageController,
  CreateMojiMessageView,
  CreateMojiMessageModel,
} from '..';
import { CreateMojiMessageStore } from '../../../domain';
import {
  CreateMojiMessageViewElements,
  EmojiSelectEvent,
} from '../create-moji-message-view';
import { when } from 'jest-when';

describe('create moji message feature', () => {
  let createMojiMessageStore: CreateMojiMessageStore;
  let elements: CreateMojiMessageViewElements;
  let mockDataLayerPush: jest.Mock;

  let emojiSelectListeners: ((e: EmojiSelectEvent) => void)[];

  const raiseEmojiSelect = (emoji: string) =>
    emojiSelectListeners.forEach((listener) =>
      listener({
        emoji,
      })
    );

  const submitMojiMessageForm = (
    message = 'Hello world',
    messageEmoji = '🌼',
    backgroundEmoji = '⚪️'
  ) => {
    elements.messageInput.value = message;
    elements.messageInput.dispatchEvent(new Event('input'));

    elements.messageEmojiPicker.click();
    raiseEmojiSelect(messageEmoji);

    elements.backgroundEmojiPicker.click();
    raiseEmojiSelect(backgroundEmoji);

    elements.createMojiMessageForm.submit();
  };

  beforeEach(() => {
    createMojiMessageStore = new CreateMojiMessageStore();

    const createMojiMessageForm = document.createElement('form');
    createMojiMessageForm.noValidate = true;
    createMojiMessageForm.id = '#messageForm';

    const messageInput = document.createElement('input');
    messageInput.required = true;

    const emojiPickerPopup = {
      addEventListener: jest.fn(),
      toggle: jest.fn(),
    };

    emojiSelectListeners = [];
    when(emojiPickerPopup.addEventListener)
      .calledWith('emoji:select', expect.anything())
      .mockImplementation((_, listener) => {
        emojiSelectListeners.push(listener);
      });

    elements = {
      createMojiMessageForm,
      messageInput,
      messageInputError: document.createElement('div'),
      messageEmojiPicker: document.createElement('button'),
      messageEmojiPickerError: document.createElement('div'),
      backgroundEmojiPicker: document.createElement('button'),
      emojiPickerPopup: emojiPickerPopup as unknown as PopupPickerController,
      mojiMessageOutputSection: document.createElement('div'),
      mojiMessageOutput: document.createElement('div'),
    };

    mockDataLayerPush = jest.fn();
    window.dataLayer = {
      push: mockDataLayerPush,
    };

    new CreateMojiMessageController(
      new CreateMojiMessageView(elements),
      new CreateMojiMessageModel(createMojiMessageStore)
    );
  });

  it('should output converted moji message', () => {
    submitMojiMessageForm();

    const mojiMessageOutput = elements.mojiMessageOutput.textContent;

    expect(mojiMessageOutput).toMatchSnapshot();
  });

  describe('validation', () => {
    it('should not convert moji message when form invalid', () => {
      submitMojiMessageForm('');

      const mojiMessageOutput = elements.mojiMessageOutput.textContent;

      expect(mojiMessageOutput).toBe('');
    });

    describe('message input', () => {
      it('should display message validation error when no message provided', () => {
        submitMojiMessageForm('');

        expect(elements.messageInputError.textContent).toBe(
          'Please enter a message.'
        );
        expect(elements.messageInput.classList).toContain('invalid');
      });

      it('should hide message validation error when message provided', () => {
        submitMojiMessageForm('');

        submitMojiMessageForm('Hello world');

        expect(elements.messageInputError.textContent).toBe('');
        expect(elements.messageInput.classList).not.toContain('invalid');
      });
    });

    describe('message emoji picker', () => {
      it('should display message emoji validation error when no emoji provided', () => {
        submitMojiMessageForm('Hello world', '');

        expect(elements.messageEmojiPickerError.textContent).toBe(
          'Please select an emoji.'
        );
        expect(elements.messageEmojiPicker.classList).toContain('invalid');
      });

      it('should hide message emoji validation error when emoji provided', () => {
        submitMojiMessageForm('Hello world', '');

        submitMojiMessageForm('Hello world', '🌼');

        expect(elements.messageEmojiPickerError.textContent).toBe('');
        expect(elements.messageEmojiPicker.classList).not.toContain('invalid');
      });
    });
  });

  describe('analytics', () => {
    it('should log convert_message event for moji message inputs', () => {
      submitMojiMessageForm();

      expect(mockDataLayerPush).toBeCalledWith({
        event: 'convert_message',
        message_length: 11,
        emoji: '🌼',
        background_emoji: '⚪️',
      });
    });
  });
});
