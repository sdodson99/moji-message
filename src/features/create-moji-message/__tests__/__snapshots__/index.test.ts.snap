// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create moji message feature should output converted moji message 1`] = `
"⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼🌼🌼🌼🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼🌼⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️🌼🌼🌼⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️⚪️🌼🌼🌼⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️🌼⚪️🌼⚪️
⚪️🌼⚪️🌼⚪️🌼⚪️
⚪️🌼🌼⚪️🌼🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️🌼🌼🌼⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️⚪️🌼🌼🌼⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼🌼🌼🌼⚪️⚪️
⚪️🌼⚪️🌼⚪️⚪️⚪️
⚪️🌼⚪️⚪️🌼⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼🌼⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🌼🌼🌼🌼⚪️⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼⚪️⚪️⚪️🌼⚪️
⚪️🌼🌼🌼🌼⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️"
`;
